docs/javascript/render-functions.md

```diff
@@ -0,0 +1,292 @@
+};
+```
+
+### 3. Objects with Multiple HTML Properties
+
+```jsx
+const MyComponent = (props, _railsContext) => {
+  return {
+    componentHtml: <div>Hello {props.name}</div>,
+    title: `<title>${props.title}</title>`,
+    metaTags: `<meta name="description" content="${props.description}" />`,
+  };
+};
+```
+
+### 4. Promises of Strings
+
+```jsx
+const MyComponent = async (props, _railsContext) => {
+  const data = await fetchData();
+  return `<div>Hello ${data.name}</div>`;
+};
+```
+
+### 5. Promises of Objects
+
+```jsx
+const MyComponent = async (props, _railsContext) => {
+  const data = await fetchData();
+  return {
+    componentHtml: `<div>Hello ${data.name}</div>`,
+  };
+};
+```
+
+### 6. Redirect Information
+
+```jsx
+const MyComponent = (props, _railsContext) => {
+  return {
+    redirectLocation: '/new-path',
+    error: null,
+    renderedHtml: null,
+  };
+};
+```
+
+## Important Limitations Observed in Tests
+
+The test file reveals some important limitations:
+
+1. **Direct String Returns Don't Work** - Returning a raw HTML string directly from a render function causes an error. The test `doesn't render html string returned directly from render function` demonstrates this.
+
+2. **Non-Promise Objects Need renderedHtml** - Non-promise objects must include a `renderedHtml` property to be valid, as shown in the test `doesn't render object without renderedHtml property`.
+
+3. **Async Functions Have Different Behavior** - Interestingly, the test `returns the object returned by async render function even if it doesn't have renderedHtml property` shows that Promise-returning functions can return objects without a `renderedHtml` property and they will still work.
+
+## Ruby Helper Functions with Render Functions
+
+### 1. react_component
+
+The `react_component` helper is used for rendering a single React component. It accepts various return types from render functions:
+
+```ruby
+# Basic usage with a component
+<%= react_component("MyComponent", props: { name: "John" }) %>
+
+# With a render function that returns JSX
+<%= react_component("MyRenderFunction", props: { name: "John" }) %>
+
+# With a render function that returns an object with renderedHtml
+<%= react_component("MyRenderFunction", props: { name: "John" }) %>
+
+# With a render function that returns a Promise
+<%= react_component("MyAsyncRenderFunction", props: { name: "John" }) %>
+```
+
+#### When to use:
+
+- When you need to render a single component
+- When you're rendering client-side only
+- When your render function returns a single HTML string
+
+#### Not suitable for:
+
+- When your render function returns an object with multiple HTML strings
+- When you need to insert content in different parts of the page
+
+### 2. react_component_hash
+
+The `react_component_hash` helper is used when your render function returns an object with multiple HTML strings. It allows you to place different parts of the rendered output in different parts of your layout.
```

Comment:
This section should also have a statement about which render function return types it accepts, like the `react_component` section above.